velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -454,6 +456,143 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
+  auto customerSelectedRowType = getRowType(kCustomer, customerColumns);
+  const auto& customerFileColumns = getFileColumnNames(kCustomer);
+  auto nationSelectedRowType = getRowType(kNation, nationColumns);
+  const auto& nationFileColumns = getFileColumnNames(kNation);
+
+  const std::string nationFilter = "n_name IN ('FRANCE', 'GERMANY')";
+  auto shipDateFilter = formatDateFilter(
+      "l_shipdate", lineitemSelectedRowType, "'1995-01-01'", "'1996-12-31'");
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId supplierScanNodeId;
+  core::PlanNodeId lineitemScanNodeId;
+  core::PlanNodeId ordersScanNodeId;
+  core::PlanNodeId customerScanNodeId;
+  core::PlanNodeId suppNationScanNodeId;
+  core::PlanNodeId custNationScanNodeId;
+
+  auto custNation =
+      PlanBuilder(planNodeIdGenerator, pool_.get())
+          .tableScan(
+              kNation, nationSelectedRowType, nationFileColumns, {nationFilter})
+          .capturePlanNodeId(custNationScanNodeId)
+          .planNode();
+
+  auto customerJoinNation =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kCustomer, customerSelectedRowType, customerFileColumns)
+          .capturePlanNodeId(customerScanNodeId)
+          .hashJoin(
+              {"c_nationkey"},
+              {"n_nationkey"},
+              custNation,
+              "",
+              {"n_name", "c_custkey"})
+          .project({"n_name as cust_nation", "c_custkey"})
+          .planNode();
+
+  auto ordersJoinCustomer =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kOrders, ordersSelectedRowType, ordersFileColumns)
+          .capturePlanNodeId(ordersScanNodeId)
+          .hashJoin(
+              {"o_custkey"},
+              {"c_custkey"},
+              customerJoinNation,
+              "",
+              {"cust_nation", "o_orderkey"})
+          .planNode();
+
+  auto suppNation =
+      PlanBuilder(planNodeIdGenerator, pool_.get())
+          .tableScan(
+              kNation, nationSelectedRowType, nationFileColumns, {nationFilter})
+          .capturePlanNodeId(suppNationScanNodeId)
+          .planNode();
+
+  auto supplierJoinNation =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kSupplier, supplierSelectedRowType, supplierFileColumns)
+          .capturePlanNodeId(supplierScanNodeId)
+          .hashJoin(
+              {"s_nationkey"},
+              {"n_nationkey"},
+              suppNation,
+              "",
+              {"n_name", "s_suppkey"})
+          .project({"n_name as supp_nation", "s_suppkey"})
+          .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kLineitem,
+              lineitemSelectedRowType,
+              lineitemFileColumns,
+              {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeId)
+          .hashJoin(
+              {"l_orderkey"},
+              {"o_orderkey"},
+              ordersJoinCustomer,
```

Comment:
Did you try joining orders after supplierJoinNation and customerJoinNation?

Reply:
Could you please elaborate? supplierJoinNation and customerJoinNation are independent and orders is joined with customer after customerJoinNation currently. This appears to be an optimal join order since joining lineitem on supplierJoinNation would be better to have at the end after the join with orders is completed.

Reply:
As per our offline discussion, this is the better plan.